### PR TITLE
action: use the most recent python from tool cache to run the GitHub Action

### PR DESCRIPTION
**Pull request by mayeut**
This allows to drop python 3.6 without losing support for ubuntu-18.04 runners c.f. https://github.com/pypa/cibuildwheel/issues/508.

~~Blocked on https://github.com/actions/setup-python/pull/411~~
~~Blocked on a new release of [actions/setup-python](https://github.com/actions/setup-python)~~
v4.1 has been released.

### CODE REVIEW

File: .github/workflows/test.yml
Comment (mayeut) [inline .github/workflows/test.yml:81]: There was no test for the GitHub Action so I duplicated the sample wheels steps to test the action.

File: action.yml
Comment (mayeut) [inline action.yml:23]: This could be reworked as a helper script with its own tests.
Syntax is Python 3.4+
Unfortunately, setup-python messes with the environment for other steps & I'd rather avoid side-effects for the rest of users workflow when using the cibuildwheel GitHub Action.
It should allow self-hosted runners to continue working out-of-the-box.

File: action.yml
Comment (mayeut) [inline action.yml:63]: This could be reworked as a helper script with its own tests.
Syntax is Python 3.4+
Unfortunately, setup-python messes with the environment for other steps & I'd rather avoid side-effects for the rest of users workflow when using the cibuildwheel GitHub Action.
It should allow self-hosted runners to continue working out-of-the-box.

File: .github/workflows/test.yml
Comment (mayeut) [reply .github/workflows/test.yml:81]: It might be worth isolating those tests a bit more (and maybe reduce them a bit as well in this case to avoid duplicate docker pulling / python installation).

File: action.yml
Comment (henryiii) [reply action.yml:63]: Can't this just be `python`? https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-running-a-python-script

Comment (henryiii) [reply action.yml:63]: This could be an action, we could put it in scikit-build/latest-python or something like that.

Comment (mayeut) [reply action.yml:63]: > or does that get Python 2 somewhere (like on macOS)?

Indeed... https://github.com/mayeut/cibuildwheel/actions/runs/2373174130

| runner | python |
|---|---|
|ubuntu-18.04|2.7.17|
|ubuntu-20.04|3.8.10|
|ubuntu-22.04|3.10.4|
|windows-2019|3.7.9|
|windows-2022|3.9.12|
|macos-10.15|2.7.18|
|macos-11|2.7.18|
|macos-12|2.7.18|

> If not, why not python3 {0}?

It works https://github.com/mayeut/cibuildwheel/actions/runs/2373224962. Good to know they've added python3 on windows !

| runner | python |
|---|---|
|ubuntu-18.04|3.6.9|
|ubuntu-20.04|3.8.10|
|ubuntu-22.04|3.10.4|
|windows-2019|3.7.9|
|windows-2022|3.9.12|
|macos-10.15|3.9.12|
|macos-11|3.9.12|
|macos-12|3.9.12|

> This could be an action, we could put it in scikit-build/latest-python or something like that.

Yes, we could. I'll try to get this into [actions/setup-python](https://github.com/actions/setup-python). (option not to mess with the environment, together with an `python-path` output - already got a PR for that last one)

Comment (mayeut) [reply action.yml]: > I'll try to get this into [actions/setup-python](https://github.com/actions/setup-python).

PR created at https://github.com/actions/setup-python/pull/411

I'll wait for feedback there to see which direction this PR will take (marking as draft for now)

Comment (mayeut) [reply action.yml]: It seems https://github.com/actions/setup-python/pull/411 will be merged in some form or another

File: action.yml
Comment (mayeut) [inline action.yml:27]: I went with a range to allow self-hosted runners not to download python 3.10 if they have a supported version in their tool cache but no python 3.10 yet.
We could however just say `3.10` if we prefer, self-hosted runners would download & install it just fine if missing from their tool cache.

Comment (joerick) [reply action.yml:27]: The range seems great to me. No download time, and a supported Python!

File: action.yml
Comment (henryiii) [inline action.yml:24]:
```suggestion
    - uses: actions/setup-python@v4
```

(4.1 supports this)

File: .github/workflows/test.yml
Comment (joerick) [reply .github/workflows/test.yml:81]: Its probably worth consolidating these Sample Build steps. I'm happy to use the Github Action one as the version we keep, given that the tests are checking the command-line invocation.

Comment (joerick) [reply .github/workflows/test.yml]: I've pushed a change to that effect.

**Review by joerick — APPROVED** [review]
LGTM!

**Comment by henryiii** [comment]
Ahh, `v4` doesn't point at `v4.1.0` yet. Why is this passing, actually?

**Comment by mayeut** [comment]
> Why is this passing, actually?

Seems that invalid inputs are just warnings... Since Python gets setup properly, there's no reason for the workflow to fail.

> Warning: Unexpected input(s) 'update-environment', valid inputs are ['python-version', 'python-version-file', 'cache', 'architecture', 'token', 'cache-dependency-path']

Let's wait for v4 being tagged again.
I'm not sure it's worth adding another workflow to test the action does not update the environment.

**Comment by henryiii** [comment]
This also should avoid getting the CMake environment variables set pointing at the global Python instance (also added in 4.1.0).

**Comment by joerick** [comment]
Shall we go back to `ubuntu-latest` runner before merging?

**Comment by henryiii** [comment]
Leaving this until after we drop 3.6 might be helpful. Okay either way.

**Comment by mayeut** [comment]
Let's get this in then.

**Pull request merged**
